### PR TITLE
Split virus and humans in EpiList

### PR DESCRIPTION
**Pull request by claireh93**
A follow up to #37, this PR does the following:

- Split `EpiList` into `HumanTypes` and `VirusTypes`. Human types hold information on how the different disease classes disperse virus across the landscape (`movement`). Virus types hold information on how well suited to the environment each virus category is (`traits`). Each hold vectors of `names` and `abundances`.

- The `EpiLandscape` is created with the correct sized arrays for `abundances.matrix` and `abundances.matrix_v` using new `counttypes()` on the HumanTypes and VirusTypes, respectively.

- The update! loops have been changed to reflect this new indexing - i.e. humanupdate! now loops through all classes in the human `abundances.matrix`.

- Tests, examples and docs updated accordingly.

- Removed old code about human movement - no longer functional.

### CODE REVIEW

File: examples/Epidemiology/SEI2HRD_example.jl [hunk 1]
```diff
@@ -44,41 +47,42 @@ area = 525_000.0km^2
...
+epilist = Simulation.SEI2HRD(traits, abun_v, abun_h, movement, param)
 rel = Gauss{eltype(epienv.habitat)}()
 
 # Create epi system with all information
 epi = EpiSystem(epilist, epienv, rel)
 
 # Add in initial infections randomly (samples weighted by population size)
 N_cells = size(epi.abundances.matrix, 2)
-samp = sample(1:N_cells, weights(1.0 .* human(epi.abundances)[2, :]), 100)
+samp = sample(1:N_cells, weights(1.0 .* human(epi.abundances)[1, :]), 100)
```
Comment (richardreeve) [inline examples/Epidemiology/SEI2HRD_example.jl]: If you have weights you can skip the 1:N_cells

File: examples/Epidemiology/SEI2HRD_example.jl [hunk 2]
```diff
@@ -44,41 +47,42 @@ area = 525_000.0km^2
...
 virus(epi.abundances)[1, samp] .= 100 # Virus pop
-human(epi.abundances)[4:5, samp] .= 10 # Inf pop
+human(epi.abundances)[3:4, samp] .= 10 # Inf pop
```
Comment (richardreeve) [inline examples/Epidemiology/SEI2HRD_example.jl]: Can't you just make some people exposed (i.e. 2) and leave it at that - isn't this overcomplicating things? And drop the virus completely.

File: examples/Epidemiology/Scotland_run.jl [hunk 3]
```diff
@@ -55,33 +56,34 @@ area = 525_000.0km^2
...
+split_pop = rand.(Multinomial.(Int.(epi.abundances.matrix[1, :]), 10))
 for j in 1:size(epi.abundances.matrix, 2)
     epi.abundances.matrix[cat_idx[:, 1], j] .= split_pop[j]
 end
 
 # Add in initial infections randomly (samples weighted by population size)
 N_cells = size(epi.abundances.matrix, 2)
-samp = sample(1:N_cells, weights(1.0 .* epi.abundances.matrix[2, :]), 100)
-epi.abundances.matrix[vcat(cat_idx[:, 4:5]...), samp] .= 10 # Inf pop
+samp = sample(1:N_cells, weights(1.0 .* epi.abundances.matrix[1, :]), 100)
```
Comment (richardreeve) [inline examples/Epidemiology/Scotland_run.jl]: Likewise, can drop the 1:N_cells

File: examples/Epidemiology/Scotland_run.jl [hunk 4]
```diff
@@ -55,33 +56,34 @@ area = 525_000.0km^2
...
+epi.abundances.matrix[vcat(cat_idx[:, 3:4]...), samp] .= 10 # Inf pop
```
Comment (richardreeve) [inline examples/Epidemiology/Scotland_run.jl]: Again, just make one group (2?) positive.

File: examples/Epidemiology/SEI2HRD_example.jl [hunk 5]
```diff
@@ -44,41 +47,42 @@ area = 525_000.0km^2
...
 # Dispersal kernels for virus and disease classes
-dispersal_dists = [1e-2km; fill(2.0km, 6); 1e-2km] # Virus disperses further than people for now
+dispersal_dists = [fill(2.0km, 6); 1e-2km] # Virus disperses further than people for now
 kernel = GaussianKernel.(dispersal_dists, 1e-10)
 movement = AlwaysMovement(kernel)
 
 # Traits for match to environment (turned off currently through param choice, i.e. virus matches environment perfectly)
-traits = GaussTrait(fill(298.0K, 8), fill(0.1K, 8))
-epilist = Simulation.SEI2HRD(traits, abun, movement, param)
+traits = GaussTrait(fill(298.0K, numvirus), fill(0.1K, numvirus))
```
Comment (richardreeve) [inline examples/Epidemiology/SEI2HRD_example.jl:59]: Why are there fits for the viruses to temperature twice (once as 0.1K?!).

File: examples/Epidemiology/UK_run.jl [hunk 6]
```diff
@@ -55,33 +56,34 @@ area = 875_000.0km^2
...
+split_pop = rand.(Multinomial.(Int.(epi.abundances.matrix[1, :]), 10))
 for j in 1:size(epi.abundances.matrix, 2)
     epi.abundances.matrix[cat_idx[:, 1], j] .= split_pop[j]
 end
 
 # Add in initial infections randomly (samples weighted by population size)
 N_cells = size(epi.abundances.matrix, 2)
-samp = sample(1:N_cells, weights(1.0 .* epi.abundances.matrix[2, :]), 100)
-epi.abundances.matrix[vcat(cat_idx[:, 4:5]...), samp] .= 10 # Inf pop
+samp = sample(1:N_cells, weights(1.0 .* epi.abundances.matrix[1, :]), 100)
```
Comment (richardreeve) [inline examples/Epidemiology/UK_run.jl]: ditto

File: src/Epidemiology/EpiGenerate.jl [hunk 7]
```diff
@@ -108,31 +93,30 @@ function classupdate!(epi::EpiSystem, timestep::Unitful.Time)
...
+                env_inf = (params.transition_virus[j, :] .* timestep .* virus(epi.abundances)[1, i]) ./ N
 
-                force_inf = (params.transition_force[j - 1, :] .* timestep .* epi.cache.virusmigration[1, i]) ./ N
+                force_inf = (params.transition_force[j, :] .* timestep .* epi.cache.virusmigration[1, i]) ./ N
 
                 # Add to transitional probabilities
-                trans_prob = (params.transition[j - 1, :] .* timestep) .+ env_inf .+  force_inf
-                trans_prob = 1.0 .- exp.(-1 .* trans_prob)
+                trans_val = (params.transition[j, :] .* timestep) .+ env_inf .+  force_inf
+                trans_prob = 1.0 .- exp.(-1 .* trans_val)
```
Comment (richardreeve) [inline src/Epidemiology/EpiGenerate.jl]: Just -trans_val

File: test/TestCases.jl [hunk 8]
```diff
@@ -52,23 +54,26 @@ function TestEpiSystem()
...
 
-    dispersal_dists = [1e-2km; fill(2.0km, 3); 1e-2km]
+    dispersal_dists = [fill(2.0km, numclasses - 1); 1e-2km]
     kernel = GaussianKernel.(dispersal_dists, 1e-10)
     movement = AlwaysMovement(kernel)
 
-    traits = GaussTrait(fill(298.0K, 5), fill(0.1K, 5))
-    epilist = SIR(traits, abun, movement, param)
+    traits = GaussTrait(fill(298.0K, numvirus), fill(0.1K, numvirus))
+    epilist = SIR(traits, abun_v, abun_h, movement, param)
```
Comment (richardreeve) [inline test/TestCases.jl:65]: Still not sure why there are two traits per virus?

File: test/canonical/test_SEI2HRD.jl [hunk 9]
```diff
@@ -15,9 +15,10 @@ save_path = (@isdefined save_path) ? save_path : pwd()
 ##
 # Set simulation parameters
 numclasses = 7
+numvirus = 1
 birth = fill(0.0/day, numclasses)
 death = fill(0.0/day, numclasses)
-virus_growth_asymp = virus_growth_symp = 1e-3/day
+virus_growth_asymp = virus_growth_symp = 1e-1/day
```
Comment (richardreeve) [inline test/canonical/test_SEI2HRD.jl]: 1e-1 is much more complicated than 0.1

File: test/canonical/test_SEI2HRD.jl [hunk 10]
```diff
@@ -68,16 +71,16 @@ kernel = GaussianKernel.(dispersal_dists, 1e-10)
 movement = AlwaysMovement(kernel)
 
 # Traits for match to environment (turned off currently through param choice, i.e. virus matches environment perfectly)
-traits = GaussTrait(fill(298.0K, numclasses), fill(0.1K, numclasses))
-epilist = Simulation.SEI2HRD(traits, abun, movement, param)
+traits = GaussTrait(fill(298.0K, numvirus), fill(0.1K, numvirus))
```
Comment (richardreeve) [inline test/canonical/test_SEI2HRD.jl:74]: ditto

File: test/canonical/test_age.jl [hunk 11]
```diff
@@ -39,7 +40,8 @@ for i in eachindex(age_cats)
     susceptible = fill(Int64.(50_000_000/age_cats[i]), age_cats[i])
     infected = fill(Int64.(10_000/age_cats[i]), age_cats[i])
     dead = fill(0, age_cats[i])
-    abun = [virus; susceptible; infected; dead]
+    abun_h = [susceptible; infected; dead]
```
Comment (richardreeve) [inline test/canonical/test_age.jl:43]: Why is this a 1xn matrix, when it was an n vector in the other tests?

**Review by richardreeve — COMMENTED** [review]
There's lots here I need to understand better, but if you want to you can just merge this without my approval.

File: examples/Epidemiology/SEI2HRD_example.jl (hunk 1 above)
Comment (richardreeve) [reply examples/Epidemiology/SEI2HRD_example.jl]: Also I think human() isn't exported, so you need to use Simulation.human(). Likewise Simulation.virus() throughout.

File: examples/Epidemiology/SEI2HRD_example.jl (hunk 2 above)
Comment (richardreeve) [reply examples/Epidemiology/SEI2HRD_example.jl]: And again Simulation.virus(), Simulation.human()

File: examples/Epidemiology/SEI2HRD_example.jl (hunk 5 above)
Comment (claireh93) [reply examples/Epidemiology/SEI2HRD_example.jl:59]: Mean and niche width for Gaussian - gave them a very narrow tolerance and exact fit to temperature for now.

File: examples/Epidemiology/SEI2HRD_example.jl (hunk 1 above)
Comment (claireh93) [reply examples/Epidemiology/SEI2HRD_example.jl]: Will export human() and virus() - they are useful functions.

File: examples/Epidemiology/SEI2HRD_example.jl (hunk 2 above)
Comment (claireh93) [reply examples/Epidemiology/SEI2HRD_example.jl]: Yes good idea - have changed.

File: test/canonical/test_SEI2HRD.jl (hunk 9 above)
Comment (claireh93) [reply test/canonical/test_SEI2HRD.jl]: Oops this was an accident from playing around with params - will switch it back.

File: test/canonical/test_age.jl (hunk 11 above)
Comment (claireh93) [reply test/canonical/test_age.jl:43]: I think this is just a vector isn't it? I'm just sticking together other vectors for susceptible, infected etc. for the number of age categories there are.

File: examples/Epidemiology/SEI2HRD_example.jl (hunk 5 above)
Comment (richardreeve) [reply examples/Epidemiology/SEI2HRD_example.jl:59]: Oh, I see!

**Review by richardreeve — APPROVED** [review]

**Comment by richardreeve** [comment]
@claireh93 I would just merge this if the tests run locally for you. I restarted the Apple CI because it had hung, but it has again now!

**Pull request merged**